### PR TITLE
fix: index_fill_ formula to support duplicate indices

### LINKED ISSUE #57006: [grad] index_fill : Incorrect gradient when index has duplicates
```python
import torch

t = torch.randn(5, 5, requires_grad=True, dtype=torch.double)
# idx = torch.tensor([3, 0]) # Works.
idx = torch.tensor([2, 2])

print(torch.unique(idx))

v = torch.tensor(-2.5122, requires_grad=True, dtype=torch.double)

def fn(v):
    return t.index_fill(0, idx, v)

torch.autograd.gradcheck(fn, (v,))
torch.autograd.gradgradcheck(fn, (v,))
```

Reason: https://github.com/pytorch/pytorch/blob/6de11c2db37d2b64debb97af40147de1581f718a/tools/autograd/derivatives.yaml#L607-L610

Formula for value doesn't consider this. And thus we get the failures.

Quick Fix (tested locally):
```patch
--- a/tools/autograd/derivatives.yaml
+++ b/tools/autograd/derivatives.yaml
@@ -606,7 +606,7 @@
 
 - name: index_fill_.int_Tensor(Tensor(a!) self, int dim, Tensor index, Tensor value) -> Tensor(a!)
   self: grad.clone().index_fill_(dim, index, 0)
-  value: grad.index_select(dim, index).sum()
+  value: grad.index_select(dim, std::get<0>(at::_unique(index))).sum()
   index: non_differentiable
 
 - name: index_put_(Tensor(a!) self, Tensor?[] indices, Tensor values, bool accumulate=False) -> Tensor(a!)
```

Questions to tackle for fixing

* Should forward error out if there are duplicate indices (my guess is this behaviour wasn't intentional) (BC breaking)?
* If we decide to keep forward as is and update the backward, there would be a perf-hit. (Is it ok to have perf hit for this niche case)

Reference to thread: https://github.com/pytorch/pytorch/issues/53237#issuecomment-827446885

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #57006

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -606,7 +606,7 @@
 
 - name: index_fill_.int_Tensor(Tensor(a!) self, int dim, Tensor index, Tensor value) -> Tensor(a!)
   self: grad.clone().index_fill_(dim, index, 0)
-  value: grad.index_select(dim, index).sum()
+  value: grad.index_select(dim, std::get<0>(at::_unique(index, /*sorted=*/false))).sum()
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:609]: What happens during the forward when the same index is passed multiple times? Is one taken arbitrarily?

Comment (ngimel) [reply tools/autograd/derivatives.yaml:609]: index_fill fills all the specified locations with a single value, so result is deterministic, but the gradient depends on how many locations in the result tensor were changed, and those locations shouldn't be double-counted.

Comment (albanD) [reply tools/autograd/derivatives.yaml:609]: Ho value is a scalar Tensor here all the time :D 

If it was the same size as index, we would have to know which value was actually written in the final output. But I was pessimistic!

**Review by albanD — APPROVED** [review]
LGTM
